torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -195,35 +177,40 @@ void DistEngine::computeDependencies(
   autogradContext->setGraphTask(std::move(graphTask));
 }
 
-void DistEngine::runEngineAndAccumulateGradients(
+std::shared_ptr<FutureVariableList> DistEngine::runEngineAndAccumulateGradients(
     const ContextPtr& autogradContext,
     const std::shared_ptr<Node>& graphRoot,
     const edge_list& outputEdges) {
-  // Kick off autograd computation with the root node and retrieve all the
-  // gradients.
-  // TODO: make this non-blocking
-  // (https://github.com/pytorch/pytorch/issues/26359)
-  ThreadsBlockedGuard guard;
-  variable_list grads = engine_
-                            .execute_with_graph_task(
-                                autogradContext->retrieveGraphTask(), graphRoot)
-                            ->wait();
-
-  // Accumulate all the gradients in the context.
-  TORCH_INTERNAL_ASSERT(grads.size() == outputEdges.size());
-  for (size_t i = 0; i < grads.size(); i++) {
-    // It is possible that the grad is not defined since a separate invocation
-    // of the autograd engine on the same node might actually compute this
-    // gradient.
-    // Also accumulate grads only for AccumulateGrad function.
-    if (grads[i].defined() &&
-        dynamic_cast<AccumulateGrad*>(outputEdges[i].function.get())) {
-      auto& variable =
-          std::static_pointer_cast<AccumulateGrad>(outputEdges[i].function)
-              ->variable;
-      autogradContext->accumulateGrad(variable, grads[i]);
-    }
-  }
+  auto futureGrads = engine_.execute_with_graph_task(
+      autogradContext->retrieveGraphTask(), graphRoot);
+
+  futureGrads->addCallback(
```

Comment:
This function does actual work in the future callback, which is a little sketchy, since the isCompleted() result isn't technically correct anymore (i.e. it can return isCompleted() == true when real work is still happening; downstream consumers like rpc_agent rely on this value being correct for short-cut codepaths)

But, this should be ok (and saves creating a Future), because it's purely an internal/private function, and you control the two use cases inside this file.

But it might be worth a short "warning" comment here, since the Future being returned here won't work in general, e.g. (you can probably make up something better than this :) )

  // Warning: the returned future is for internal use only, since it does
  // processing in the callback, hence isCompleted() is going to be set earlier
  // than the actual logical completion of the future.

Or maybe just create a separate future, though that's not strictly necessary.

Reply:
Looks great, thanks for separating this out!

(btw, I do agree with the decision in Future<> to set completed=true at the beginning of markCompleted(), since I don't think we want to deal with callbacks fiddling with the future, and all the odd corner cases that could happen there as a result :) )